lib/OpenQA/Worker/Cache.pm

```diff
@@ -263,106 +236,46 @@ sub get_asset {
     return $asset;
 }
 
-sub toggle_asset_lock {
-    my $self = shift;
-    my ($asset, $toggle) = @_;
-
-    my $dbh = $self->dbh;
-
-    my $check = 1 - $toggle;
-    my $sql
-      = "UPDATE assets set downloading = ?, filename = ?, last_use = strftime('%s','now') where filename = ? and downloading = ?";
-    my $res;
-    eval {
-        $dbh->do("BEGIN EXCLUSIVE");
-        $res = $dbh->prepare($sql)->execute($toggle, $asset, $asset, $check);
-        $dbh->commit or die $dbh->errstr;
-    };
-
-    if ($@) {
-        log_error "toggle_asset_lock: Rolling back $@";
-        eval { $dbh->rollback };
-        log_error "Rolling back failed: $@" if $@;
-    }
-
-    return 0 if !defined $res || $res == 0 || $@;
-    return 1 if defined $res && $res == 1;
-}
-
 sub _asset {
     my ($self, $asset) = @_;
+    my $result = $self->dbh->db->select('assets', [qw(etag size last_use)], {filename => $asset})->hashes;
 
-    my $sql = "SELECT downloading, etag from assets where filename = ?";
-    $self->dbh->prepare($sql);
-    my $result = $self->dbh->selectrow_hashref($sql, undef, $asset);
-
-    return {} unless $result;
-    return $result;
-}
-
-sub try_lock_asset {
-    my ($self, $asset) = @_;
-    my $sth;
-    my $sql;
-    my $lock_granted = 0;
-    my $result;
-
-    my $dbh = $self->dbh;
-
-    eval {
-        $lock_granted = 1 if $self->toggle_asset_lock($asset, 1);
-        $result = $self->_asset($asset);
-        log_info "CACHE: Being downloaded by another worker, sleeping."
-          if exists $result->{downloading} && $result->{downloading} == 1 && $lock_granted == 0;
-    };
-    log_error "try_lock_asset: Failed $@" if $@;
-
-    return $result if $lock_granted;
-    return !!0;
+    return {} if $result->size == 0 || $@;
+    return $result->first;
 }
 
 sub track_asset {
     my ($self, $asset) = @_;
 
-    my $dbh = $self->dbh;
     my $res;
-    my $sql
-      = "INSERT OR IGNORE INTO assets (downloading,filename, size, last_use) VALUES (0, ?, 0,  strftime('%s','now'));";
+    my $sql = "INSERT OR IGNORE INTO assets (filename, size, last_use) VALUES (?, 0,  strftime('%s','now'));";
 
     eval {
-        $dbh->do("BEGIN EXCLUSIVE");
-        $res = $dbh->prepare($sql)->execute($asset) or die $dbh->errstr;
-        $dbh->commit or die $dbh->errstr;
+        my $tx = $self->dbh->db->begin('exclusive');
+        $res = $self->dbh->db->query($sql, $asset)->arrays;
+        $tx->commit;
```

Comment:
Another case where you need to hold on to the db handle instead of calling `$self->dbh->db` repeatedly.